app-libs/parentchain-interface/src/target_b/event_handler.rs

```diff
@@ -63,21 +82,38 @@ where
 		events: impl FilterEvents,
 		vault_account: &AccountId,
 	) -> Result<(), Error> {
-		let filter_events = events.get_transfer_events();
 		trace!(
-			"[TargetB] filtering transfer events to shard vault account: {}",
+			"[TargetB] filtering balance transfer events to shard vault account: {}",
 			hex_encode(vault_account.encode().as_slice())
 		);
+		let filter_events = events.get_events::<BalanceTransfer>();
 		if let Ok(events) = filter_events {
 			events
-				.iter()
-				.filter(|&event| event.to == *vault_account)
-				.try_for_each(|event| {
-					info!("[TargetB] found transfer event to shard vault account: {} will shield to {}", event.amount, hex_encode(event.from.encode().as_ref()));
-					Self::shield_funds(executor, &event.from, event.amount)
-				})
-				.map_err(|_| ParentchainError::ShieldFundsFailure)?;
+                .iter()
+                .filter(|&event| event.to == *vault_account)
+                .try_for_each(|event| {
+                    info!("[TargetB] found balance transfer event to shard vault account: {} will shield to {}", event.amount, hex_encode(event.from.encode().as_ref()));
+                    Self::shield_funds(executor, &event.from, event.amount, None)
+                })
+                .map_err(|_| ParentchainError::ShieldFundsFailure)?;
 		}
+		trace!(
+			"[TargetB] filtering foreign assets transferred events to shard vault account: {}",
+			hex_encode(vault_account.encode().as_slice())
+		);
+		let filter_asset_events = events.get_events::<ForeignAssetsTransferred>();
+		if let Ok(events) = filter_asset_events {
+			events
+                .iter()
+                .filter(|&event| event.to == *vault_account)
+                .try_for_each(|event| {
+                    let stf_asset_id = AssetId::from_location(&event.asset_id);
+                    info!("[TargetB] found foreign assets ({:?}) transferred event to shard vault account: {} will shield to {}", stf_asset_id, event.amount, hex_encode(event.from.encode().as_ref()));
```

Comment:
This seems like some duplicate code here.

Reply:
True, we could use a generic method here, but the necessary branching would likely result in equal lines of code. For now I'd consider this acceptable ;-)